In [1]:
def numerical_derivative(f, x, h=1e-7):
    return (f(x + h) - f(x - h)) / (2 * h)

def f(x):
    return x ** 2

for x in [-2, -1, 0, 1, 2]:
    numerical = numerical_derivative(f, x)
    analytical = 2 * x
    print(f"x={x:2d}  f'(x) 数值={numerical:.6f}  解析={analytical:.1f}")

x=-2  f'(x) 数值=-4.000000  解析=-4.0
x=-1  f'(x) 数值=-2.000000  解析=-2.0
x= 0  f'(x) 数值=0.000000  解析=0.0
x= 1  f'(x) 数值=2.000000  解析=2.0
x= 2  f'(x) 数值=4.000000  解析=4.0


In [2]:
def numerical_gradient(f, point, h=1e-7):
    gradient = []
    for i in range(len(point)):
        point_plus = list(point)
        point_minus = list(point)
        point_plus[i] += h
        point_minus[i] -= h
        partial = (f(point_plus) - f(point_minus)) / (2 * h)
        gradient.append(partial)
    return gradient

def f_multi(point):
    x, y = point
    return x**2 + 3*x*y + y**2

grad = numerical_gradient(f_multi, [1.0, 2.0])
print(f"在 (1,2) 处的数值梯度: {[f'{g:.4f}' for g in grad]}")
print(f"在 (1,2) 处的解析梯度: [2*1+3*2, 3*1+2*2] = [{2*1+3*2}, {3*1+2*2}]")

在 (1,2) 处的数值梯度: ['8.0000', '7.0000']
在 (1,2) 处的解析梯度: [2*1+3*2, 3*1+2*2] = [8, 7]


In [3]:
x = 5.0
lr = 0.1
for step in range(20):
    grad = 2 * x
    x = x - lr * grad
    print(f"step {step:2d}  x={x:8.4f}  f(x)={x**2:10.6f}")

step  0  x=  4.0000  f(x)= 16.000000
step  1  x=  3.2000  f(x)= 10.240000
step  2  x=  2.5600  f(x)=  6.553600
step  3  x=  2.0480  f(x)=  4.194304
step  4  x=  1.6384  f(x)=  2.684355
step  5  x=  1.3107  f(x)=  1.717987
step  6  x=  1.0486  f(x)=  1.099512
step  7  x=  0.8389  f(x)=  0.703687
step  8  x=  0.6711  f(x)=  0.450360
step  9  x=  0.5369  f(x)=  0.288230
step 10  x=  0.4295  f(x)=  0.184467
step 11  x=  0.3436  f(x)=  0.118059
step 12  x=  0.2749  f(x)=  0.075558
step 13  x=  0.2199  f(x)=  0.048357
step 14  x=  0.1759  f(x)=  0.030949
step 15  x=  0.1407  f(x)=  0.019807
step 16  x=  0.1126  f(x)=  0.012677
step 17  x=  0.0901  f(x)=  0.008113
step 18  x=  0.0721  f(x)=  0.005192
step 19  x=  0.0576  f(x)=  0.003323


In [4]:
def f_2d(point):
    x, y = point
    return x**2 + y**2

point = [4.0, 3.0]
lr = 0.1
for step in range(30):
    grad = numerical_gradient(f_2d, point)
    point = [p - lr * g for p, g in zip(point, grad)]
    loss = f_2d(point)
    if step % 5 == 0 or step == 29:
        print(f"step {step:2d}  point=({point[0]:7.4f}, {point[1]:7.4f})  f={loss:.6f}")

step  0  point=( 3.2000,  2.4000)  f=16.000000
step  5  point=( 1.0486,  0.7864)  f=1.717987
step 10  point=( 0.3436,  0.2577)  f=0.184467
step 15  point=( 0.1126,  0.0844)  f=0.019807
step 20  point=( 0.0369,  0.0277)  f=0.002127
step 25  point=( 0.0121,  0.0091)  f=0.000228
step 29  point=( 0.0050,  0.0037)  f=0.000038


In [5]:
import math

test_functions = [
    ("x^2",      lambda x: x**2,          lambda x: 2*x),
    ("x^3",      lambda x: x**3,          lambda x: 3*x**2),
    ("sin(x)",   lambda x: math.sin(x),   lambda x: math.cos(x)),
    ("e^x",      lambda x: math.exp(x),   lambda x: math.exp(x)),
    ("1/x",      lambda x: 1/x,           lambda x: -1/x**2),
]

x = 2.0
print(f"{'函数':<12} {'数值':>12} {'解析':>12} {'误差':>12}")
print("-" * 50)
for name, f, df in test_functions:
    num = numerical_derivative(f, x)
    ana = df(x)
    err = abs(num - ana)
    print(f"{name:<12} {num:12.6f} {ana:12.6f} {err:12.2e}")

函数                     数值           解析           误差
--------------------------------------------------
x^2              4.000000     4.000000     4.33e-09
x^3             12.000000    12.000000     6.32e-09
sin(x)          -0.416147    -0.416147     1.12e-10
e^x              7.389056     7.389056     7.50e-09
1/x             -0.250000    -0.250000     1.32e-10


In [6]:
def hessian_2d(f, x, y, h=1e-5):
    fxx = (f(x + h, y) - 2 * f(x, y) + f(x - h, y)) / (h ** 2)
    fyy = (f(x, y + h) - 2 * f(x, y) + f(x, y - h)) / (h ** 2)
    fxy = (f(x + h, y + h) - f(x + h, y - h) - f(x - h, y + h) + f(x - h, y - h)) / (4 * h ** 2)
    return [[fxx, fxy], [fxy, fyy]]

def saddle(x, y):
    return x ** 2 - y ** 2

def bowl(x, y):
    return x ** 2 + y ** 2

H_saddle = hessian_2d(saddle, 0.0, 0.0)
H_bowl = hessian_2d(bowl, 0.0, 0.0)
print(f"鞍点海森矩阵: {H_saddle}")  # [[2, 0], [0, -2]] -- 符号混合
print(f"碗状海森矩阵:   {H_bowl}")    # [[2, 0], [0, 2]]  -- 都为正

鞍点海森矩阵: [[2.0, 0.0], [0.0, -2.0]]
碗状海森矩阵:   [[2.0, 0.0], [0.0, 2.0]]


In [7]:
import math

def taylor_approx(f, f_prime, f_double_prime, x0, h, order=2):
    result = f(x0)
    if order >= 1:
        result += f_prime(x0) * h
    if order >= 2:
        result += 0.5 * f_double_prime(x0) * h ** 2
    return result

x0 = 0.0
for h in [0.1, 0.5, 1.0, 2.0]:
    true_val = math.sin(h)
    t1 = taylor_approx(math.sin, math.cos, lambda x: -math.sin(x), x0, h, order=1)
    t2 = taylor_approx(math.sin, math.cos, lambda x: -math.sin(x), x0, h, order=2)
    print(f"h={h:.1f}  sin(h)={true_val:.4f}  一阶={t1:.4f}  二阶={t2:.4f}")

h=0.1  sin(h)=0.0998  一阶=0.1000  二阶=0.1000
h=0.5  sin(h)=0.4794  一阶=0.5000  二阶=0.5000
h=1.0  sin(h)=0.8415  一阶=1.0000  二阶=1.0000
h=2.0  sin(h)=0.9093  一阶=2.0000  二阶=2.0000


In [9]:
import random

random.seed(42)

w = random.gauss(0, 1)
b = random.gauss(0, 1)
lr = 0.01

xs = [1.0, 2.0, 3.0, 4.0, 5.0]
ys = [3.0, 5.0, 7.0, 9.0, 11.0]

for epoch in range(1200):
    total_loss = 0
    dw = 0
    db = 0
    for x, y in zip(xs, ys):
        pred = w * x + b
        error = pred - y
        total_loss += error ** 2
        dw += 2 * error * x
        db += 2 * error
    dw /= len(xs)
    db /= len(xs)
    total_loss /= len(xs)
    w -= lr * dw
    b -= lr * db
    if epoch % 40 == 0 or epoch == 199:
        print(f"epoch {epoch:3d}  w={w:.4f}  b={b:.4f}  loss={total_loss:.6f}")

print(f"\n学习结果: y = {w:.2f}x + {b:.2f}")
print(f"真实关系:  y = 2x + 1")

epoch   0  w=0.3980  b=-0.0208  loss=67.032927
epoch  40  w=2.1296  b=0.5319  loss=0.040145
epoch  80  w=2.1132  b=0.5912  loss=0.030618
epoch 120  w=2.0989  b=0.6430  loss=0.023351
epoch 160  w=2.0863  b=0.6883  loss=0.017809
epoch 199  w=2.0757  b=0.7268  loss=0.013675
epoch 200  w=2.0754  b=0.7278  loss=0.013582
epoch 240  w=2.0659  b=0.7622  loss=0.010359
epoch 280  w=2.0575  b=0.7924  loss=0.007900
epoch 320  w=2.0502  b=0.8187  loss=0.006025
epoch 360  w=2.0439  b=0.8416  loss=0.004595
epoch 400  w=2.0383  b=0.8617  loss=0.003505
epoch 440  w=2.0335  b=0.8792  loss=0.002673
epoch 480  w=2.0292  b=0.8945  loss=0.002039
epoch 520  w=2.0255  b=0.9079  loss=0.001555
epoch 560  w=2.0223  b=0.9196  loss=0.001186
epoch 600  w=2.0195  b=0.9298  loss=0.000904
epoch 640  w=2.0170  b=0.9387  loss=0.000690
epoch 680  w=2.0148  b=0.9464  loss=0.000526
epoch 720  w=2.0130  b=0.9532  loss=0.000401
epoch 760  w=2.0113  b=0.9591  loss=0.000306
epoch 800  w=2.0099  b=0.9643  loss=0.000233
epoch 84

## 上述的结果有什么意义呢？

这个结果展示了**梯度下降从随机初始化开始，逐步逼近真实参数的过程**，但也暴露了有限迭代次数和固定学习率带来的**近似性**。

### 具体意义：

1. **梯度下降确实有效**  
   初始时 `w=0.398, b=-0.0208`，损失高达 67。经过 200 轮更新，损失降到了 0.0137，`w` 接近 2，`b` 从负值上升到 0.73。说明模型学到了数据中的基本线性趋势：`y ≈ 2x`。

2. **为什么没有精确学到 `y = 2x + 1`？**  
   - **迭代次数不足**：200 个 epoch 对这个问题足够吗？如果继续训练，`w` 和 `b` 会进一步接近 2 和 1，损失会趋近 0。当前结果只是“中途快照”。  
   - **学习率偏小**：`lr=0.01` 步长较小，收敛慢。增大学习率或使用自适应方法（如 Adam）可以更快接近真值。  
   - **数据无噪声，理论上可完美拟合**：因为数据完全来自 `y = 2x + 1`，没有噪声。所以损失最终可以降到 0。目前没到 0 只是优化还没跑够。

3. **实际启示**  
   - 训练深度学习模型时，**你很少能得到“完美”的参数**（真值不可知），只能得到使损失足够小的近似解。  
   - 监控损失曲线的下降趋势比盯着参数绝对值更重要。这里损失从 67 → 0.0137，说明方向正确。  
   - 如果损失不再下降，可以尝试：增大学习率、增加迭代次数、调整初始化、加入动量等。

### 一句话总结
> **梯度下降从瞎猜开始，每一步都让模型更接近真相，但有限的时间和步长让它停在了一个“足够好”的近似解——这正是所有神经网络训练的缩影。**